pkg/portal/cluster/nodes.go

```diff
@@ -0,0 +1,135 @@
+package cluster
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type NodeConditions struct {
+	Type               string `json:"type"`
+	Status             string `json:"status"`
+	LastHeartbeatTime  string `json:"lastHeartbeatTime,omitempty"`
+	LastTransitionTime string `json:"lastTransitionTime,omitempty"`
+	Reason             string `json:"reason,omitempty"`
+	Message            string `json:"message,omitempty"`
+}
+
+type Taint struct {
+	Key       string `json:"key"`
+	Value     string `json:"value,omitempty"`
+	Effect    string `json:"effect"`
+	TimeAdded string `json:"timeAdded,omitempty"`
+}
+
+type MachineResources struct {
+	CPU           string
+	StorageVolume string
+	Memory        string
+	Pods          string
+}
+
+type Volume struct {
+	Name string
+	Path string
+}
+
+type NodeInformation struct {
+	Name        string            `json:"name"`
+	CreatedTime string            `json:"createdTime"`
+	Capacity    MachineResources  `json:"capacity"`
+	Volumes     []Volume          `json:"volumes"`
+	Allocatable MachineResources  `json:"allocatable"`
+	Taints      []Taint           `json:"taints"`
+	Conditions  []NodeConditions  `json:"conditions"`
+	Labels      map[string]string `json:"labels"`
+	Annotations map[string]string `json:"annotations"`
+}
+
+type NodeListInformation struct {
+	Nodes []NodeInformation `json:"nodes"`
+}
+
+func NodesFromNodeList(nodes *corev1.NodeList) *NodeListInformation {
+	final := &NodeListInformation{
+		Nodes: make([]NodeInformation, 0, len(nodes.Items)),
+	}
+
+	for _, node := range nodes.Items {
+		taints := []Taint{}
+		// TODO: Add Null fields seperately!
```

Comment:
@s-amann Suggestion was even better as it applied DRY, but if it can't be applied, no more debate.

Regarding a single long function, IMHO it is always cleaner/better to have a short function that gives you an overall idea about what the code does without going further into the details. I think good encapsulation/high cohesion starts with having short functions. 

The clear advantage I see with the multiple functions I suggested is that you can understand what "MachinesFromMachineList" does reading just 18 LOC (the function itself) and later you can go read each of the narrow functions it calls to understand the details. I did "comment PR" instead of "suggest changes" as there is nothing wrong with the code and this is just a personal suggestion to improve it a bit (from my personal point of view). 

There is an interesting talk about [Clean Code functions](https://www.youtube.com/watch?v=QvrNwfuWbCQ) from Uncle Bob.

In the end the decision is up to you 🙂

Reply:
I need to ask @Ellis about what he wants to do in this. I will get back to you on this.

Reply:
I believe the TODO here was addressed by the expression below so it can be removed now

Reply:
@s-amann  Well this has been addressed from the very next line .. this comment can be removed, I will remove it.